### PR TITLE
Improvements to grover runtime

### LINKED ISSUE #3511: Improving grover training time
The grover model takes a long time to train. On a dataset with ~900 datapoints (delaney train), one epoch takes 
about 30 seconds.

In a single iteration, approximately 2.19 sec is consumed in extracting grover graph attributes
https://github.com/deepchem/deepchem/blob/60d1fd3be16f657dd3c6cb9b0bc56a1144d5de5d/deepchem/models/torch_models/grover.py#L591-L592

and only 0.50 sec in the `forward` method for pretraining models (the numbers 2.19 and 0.5 are averaged over 30 iterations).

If we can prepare batched data before training (extracting graph attributes), it can speedup grover model 
training ~5x times.

### PR DESCRIPTION
**Pull request by arunppsg**
## Description

Fix #3511

Improved grover runtime.

Summary of changes:
- updated grover featurizer to pre-compute grover graph features during dataset featurization phase
- added `BatchGroverGraph` to handle edge cases in batching of grover molecular graph
- minor changes to grover torch model

Testing setup:

Accelerator: Apple M1 GPU
Dataset: Delaney (1127 data points)
Other parameters: Batch size 100, number of epochs: 10

Testing results:

 | Approach | Total training time | Avg. time for batch preperation | Avg. time for extracting attributes | Avg. time on forward pass | Avg. time for an iteration |
 | --- | ---- | ---- | ---- | ---- | ---- |
 | New | 454 sec | 0.007 sec | 2.68e-06 sec | 1.043 sec | 1.05 sec |
 | Old | 138 sec | 0.002 sec | 2.65 sec | 1.026 sec | 3.682 sec

Total training time on old approach: 454 seconds
Total training time on new approach: 138 seconds (~3.2x speedup)

Average time corresponds to time taken for an iteration.

## Type of change

Please check the option that is related to your PR.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/utils/grover.py [hunk 1]
```diff
@@ -1,13 +1,94 @@
...
 
 try:
     import torch
 except ModuleNotFoundError:
     pass
 
+from deepchem.feat.molecule_featurizers.dmpnn_featurizer import GraphConvConstants
+
+
+class BatchGroverGraph:
```
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/utils/grover.py:14]: Also, can you add a paragraph explaining how the batching method works?

File: deepchem/utils/grover.py [hunk 2]
```diff
@@ -1,13 +1,94 @@
...
+from deepchem.feat.graph_data import BatchGraphData, GraphData
 
 try:
     import torch
 except ModuleNotFoundError:
     pass
 
+from deepchem.feat.molecule_featurizers.dmpnn_featurizer import GraphConvConstants
+
+
```
Comment (rbharath) [inline deepchem/utils/grover.py:13]: Let's add a simple unit test for `BatchGroverGraph` as well

File: deepchem/feat/molecule_featurizers/grover_featurizer.py [hunk 3]
```diff
@@ -143,9 +152,26 @@ def _make_mol_graph(self, mol: RDKitMol) -> GraphData:
 
                 edge_index.extend([[a1, a2], [a2, a1]])
 
+                b1 = n_bonds
```
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/grover_featurizer.py:155]: Can you add some comments here explaining what these lines do? It's a bit confusing to read

**Review by rbharath — COMMENTED** [review]
Some requests for additional comments and tests

File: deepchem/utils/grover.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (arunppsg) [reply deepchem/utils/grover.py:14]: added them

File: deepchem/utils/grover.py (hunk 2 above)
Comment (arunppsg) [reply deepchem/utils/grover.py:13]: added

File: deepchem/feat/molecule_featurizers/grover_featurizer.py (hunk 3 above)
Comment (arunppsg) [reply deepchem/feat/molecule_featurizers/grover_featurizer.py:155]: added

**Review by rbharath — APPROVED** [review]
LGTM

**Pull request merged**